mbed_tools_ci/utils/git_helpers.py

```diff
@@ -462,3 +554,76 @@ def is_current_branch_feature(self) -> bool:
         is_beta = current_branch == self.get_beta_branch()
         is_release = self.is_release_branch(current_branch)
         return not (is_master or is_beta or is_release)
+
+    @property
+    def uncommitted_changes(self) -> List[Path]:
+        """Gets list of uncommitted files.
+
+        Returns:
+             list of uncommitted
+        """
+        status = self.repo.git.status(
+            porcelain=True,
+            untracked_files=True
+
+        )
+        if not status:
+            return []
+
+        return [
+            Path(self.root).joinpath(
+                line.strip().split(' ')[-1]) for line in status.splitlines()
+        ]
+
+    def apply_uncommitted_changes(self, other_repo: Union[Path, 'GitWrapper']) -> None:
+        """Applies the uncommitted changes found in current repository to another.
+
+        Args:
+            other_repo: repository or path to repository to apply changes to
+        """
+        dest_root = other_repo if isinstance(other_repo,
+                                             Path) else other_repo.root
+        for f in self.uncommitted_changes:
+            destination = dest_root.joinpath(f.relative_to(self.root))
+            if f.exists():
+                logger.info(f"Applying change in {f} to {destination}")
+                if not destination.parent.exists():
+                    os.makedirs(str(destination.parent), exist_ok=True)
+                shutil.copy2(src=str(f), dst=str(destination))
+            else:
+                logger.info(f"Removing {destination}")
+                if destination.exists():
+                    destination.unlink()
+
+
+class GitTempClone:
+    """Context manager providing a temporary cloned repository."""
+
+    def __init__(self, current_branch_name: Optional[str] = None,
```

Comment:
1. This is never called with `repository_to_clone` other than in test. 
Also, I would avoid the pattern of magic defaults (cause `Optional[]` stinks!) and wrap those use cases in specific functions or classes. Otherwise, I think this is walking on the edge of single responsibility principle.
My suggestion:
```
class GitTempClone:
    def __init__(self, repository_to_clone: GitWrapper):  # this always wants repository_to_clone
        self._clone = repository_to_clone

class ProjectRootGitTempClone(GitTempClone):
    def __init__(self):
        git = GitWrapper(configuration.get_value(ConfigurationVariable.PROJECT_ROOT))
        super().__init__(repository_to_clone=git) 
```

2. `temporary_directory` is never passed in, why have it as argument?

3. I think the branching should not happen here. Same with `apply_uncomitted_changes` (what does it even do?). Just do a clone.

4. Finally, I think you're better off using more succinct:
```
@contextlib.contextmanager
def git_temp_clone(repository: GitWrapper):
    with TemporaryDirectory() as temp_dir:
        clone = repository.cone(temp_dir.path)
        yield clone

@contextlib.contextmanager
def git_temp_clone_of_project_root():
    git = GitWrapper(configuration.get_value(ConfigurationVariable.PROJECT_ROOT))
    with git_temp_clone(git) as clone:
        yield clone
```

Reply:
1 and 2 agree
3 the thing only works because of apply_uncomitted_changes. changes happen in the initial repo and the clone is created to commit them back to the remote repo (this is what tag and release does) so the clone needs to get those changes

Reply:
Re 3.
If that's the case we should clone first and then do the changes in the cloned repo only. At the moment _random_ external state is leaked into this class. Ideally we should be able to use it on its own w/o all the burden of how our CI works. :)

I do understand if that's not possible.

Reply:
the utility is to mess up with a git repo which is not the primary repository. other scripts or CI steps which have no knowledge of git or source control should not be aware of this mechanism and of the existence of a clone. this is why this is done this way and what I absolutely don't want is to put the burden of git management to other scripts: it is not their concerns.

Reply:
I'm not saying we should put this in our script files. I'm saying this does a bit much by assuming the state of the "cloned" repo. The fact that there's this global state (git changes) we have to carry around is not great, as you have no control over it and need to replicate it. In the ideal world you shouldn't.

I.e., the release process should be something like:
```
with GitTempClone(git=GitWrapper(PROJECT_ROOT)) as git_clone:
    # git and git_clone should be CLEAN at this point, why isn't it and we need to `apply_uncommited_changes`?
    update_documentation(git_clone)
    git.add('_version', 'changelog')
    git.commit('release!')
    git.tag()
    git.push()
```